File: backend/app/schemas/wallet.py

from pydantic import BaseModel, Field

from app.schemas.base import MoneyField, PositiveMoney


class WalletResponse(BaseModel):
    balance: MoneyField
    locked_balance: MoneyField
    available_balance: MoneyField
    currency: str

    model_config = {"from_attributes": True}


class DepositRequest(BaseModel):
    amount: PositiveMoney


class DepositResponse(BaseModel):
    client_secret: str
    payment_intent_id: str
    amount: MoneyField
    currency: str


class WithdrawRequest(BaseModel):
    amount: PositiveMoney
    # Idempotency key — client-generated, prevents double-withdrawal on retry
    idempotency_key: str | None = Field(None, max_length=64)


class TransactionResponse(BaseModel):
    id: str
    type: str
    amount: MoneyField
    balance_after: MoneyField
    status: str
    created_at: str

    model_config = {"from_attributes": True}

File: frontend/apps/web/lib/schemas/wallet.ts
import { z } from "zod"

// Accept both string (form/API) and number (parseFloat result) — backend Decimal coerces either
const positiveMoney = z.union([z.string(), z.number()])

export const depositSchema = z.object({
  amount: positiveMoney,
})

export const withdrawSchema = z.object({
  amount: positiveMoney,
  idempotency_key: z.string().max(64).optional(),
})

export const depositResponseSchema = z.object({
  client_secret: z.string(),
  payment_intent_id: z.string(),
  amount: z.string(),
  currency: z.string(),
})

export const walletResponseSchema = z.object({
  balance: z.string(),
  locked_balance: z.string(),
  available_balance: z.string(),
  currency: z.string(),
})

export const transactionResponseSchema = z.object({
  id: z.string(),
  type: z.string(),
  amount: z.string(),
  balance_after: z.string(),
  status: z.string(),
  created_at: z.string(),
})

export type DepositInput = z.infer<typeof depositSchema>
export type WithdrawInput = z.infer<typeof withdrawSchema>
export type DepositResponse = z.infer<typeof depositResponseSchema>
export type WalletResponse = z.infer<typeof walletResponseSchema>
export type TransactionResponse = z.infer<typeof transactionResponseSchema>
